Improve Controller Script Error Reporting

Holzhaus opened the pull request:
If a controller script contains syntax errors or cannot be opened (e.g. due to permission errors), we don't show a popup when trying to initalize it. The user will often only notice it when pressing a button when Mixxx fails to invoke `MyController.someFunction`, not instantly when loading the script. This could even happen in the middle of a set, e.g. when using something like a Korg nanoKONTROL for FX only. This behaviour also makes it unnecessarily hard for mapping developers to debug issues, because the actual error message is not shown, just a meaningless error that was caused by the failed script load.

This enables the error reporting on failed script load, and also uninitializes the script engine in that case. This means that a single error dialog will be shown when the script load fails, but subsequent button presses on the controller will do nothing and not open a popup every time. Auto-reloading the script when it changes (or the permissions are changed) will still work fine.

I also modified the error messages a bit. All error dialogs should now show the controller name (in case multiple controllers are connected). Script filenames and error messages from the OS or the script engine were moved into the "Details" section of the error messages, because in most cases they are not interesting to a regular user (failures when opening the file is a special case, but I moved the filename to the details part anyway for consistency).

Holzhaus committed:
controllers/controllerengine: Improve error messages for mappings

Holzhaus committed:
controllers/controllerengine: Enable errors popups in all cases

Holzhaus committed:
controllers/controllerengine: Add missing nullptr checks

Holzhaus committed:
controllers/controllerengine: Move uninitialize engine code into method

Holzhaus committed:
controllers/controllerengine: Uninitalize if script load failed

Holzhaus commented:
I don't know if this should go into 2.3 but IMHO the current behaviour is confusing and a major annoyance when writing controller mappings.

ferranpujolcamins commented:
This is already solved in https://github.com/ferranpujolcamins/mixxx/tree/migrate-to-QJSEngine

Which we'll release with 2.4 and we'll merge as soon as the 2.3 branch is cut.

Please have a look at my branch and rebase there if you think this is still relevant.

Holzhaus commented:
Nice!

 I think we should merge this PR for 2.3 and just replace it with your code in 2.4. It will be at least half a year between the 2.3 and 2.4 releases and I think controller mapping creators would really appreciate getting notified where exactly the syntax error is.

Be-ing commented:
I'm okay with merging this for 2.3 if you agree to take care of merge conflicts with #1795, @Holzhaus. I will probably not review this before the initial 2.3 beta release.

Holzhaus commented:
Solving merge conflicts is not exactly my favorite activity, but someone gotta do it eventually. I'll do my best.

Holzhaus committed:
Merge branch 'master' of github.com:mixxxdj/mixxx into script-error-reporting

Holzhaus committed:
controllers/controllerengine: Remove duplicated code and fix popup logic

Holzhaus committed:
controllers/controllerengine: Always show the backtrace on script errors

Previously the backtrace was only added to the error details if
the `--controllerDebug` flag was passed to Mixxx. Since the error
details are only displayed if the user explicitly uncollapsed the
"Details" expander and non-technical users won't be able to make use of
the engine's error message and line number anyway, we can just show it
unconditionally. This might even result in more helpful bug reports.

Holzhaus committed:
controllers/controllerengine: Fix error message text

Holzhaus committed:
controllers/controllerengine: Further improve error dialog details

Holzhaus committed:
controllers/controllerengine: Use monospace font for script error details

Holzhaus commented:
Ok, this is now in a good state IMHO. This should vastly improve DX for controller mapping developers. Please review.

Holzhaus commented:
Anything else to do here?

Be-ing reviewed (changes requested):
When I save a script with a syntax error, then fix the error and save again, the script is not reloaded until I restart Mixxx. I see this in the log:
```
Critical [Controller]: DEBUG ASSERT: "m_pEngine != nullptr" in function QScriptValue ControllerEngine::wrapFunctionCode(const QString &, int) at /home/be/sw/mixxx/src/controllers/controllerengine.cpp:114
```

Holzhaus committed:
controllers/controllerengine: Fix wrong debug assertion

Holzhaus committed:
controllers/controllerengine: Remove unusued getErrors method

Holzhaus committed:
controllers/controllerengine: Clear script errors upon reinitialization

Holzhaus committed:
controllers/controllerengine: Make errors during script init fatal

Holzhaus committed:
controllers/controllerengine: Removed unused initialized signal

Holzhaus committed:
controllers/controllerengine: Do not translate script error messages

Holzhaus committed:
controllers/controllerengine: Improve file open error message

Holzhaus committed:
controllers/controllerengine: Stop engine if controller init fails

Holzhaus committed:
controllers/controllerengine: Improve syntax check error message

Holzhaus commented on src/controllers/controllerengine.cpp:
```diff
@@ -82,6 +76,10 @@ Output:  -
 -------- ------------------------------------------------------ */
 void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefixes,
                                              const QString& function, QScriptValueList args) {
+    if (m_pEngine == nullptr) {
```
No, this is called from outside. However, I can check if we can improve the invocation as well, but that would probably cause even more merge conflicts.

Holzhaus commented on src/controllers/controllerengine.cpp:
```diff
@@ -311,31 +332,41 @@ bool ControllerEngine::evaluate(const QString& filepath) {
         case (QScriptSyntaxCheckResult::Valid): break;
         case (QScriptSyntaxCheckResult::Intermediate):
-            error = "Incomplete code";
+            error = tr("Incomplete code");
```
Ok, I removed it and added some comments that explain why this is not translated.

Holzhaus commented:
> When I save a script with a syntax error, then fix the error and save again, the script is not reloaded until I restart Mixxx. I see this in the log:
> 
> ```
> Critical [Controller]: DEBUG ASSERT: "m_pEngine != nullptr" in function QScriptValue ControllerEngine::wrapFunctionCode(const QString &, int) at /home/be/sw/mixxx/src/controllers/controllerengine.cpp:114
> ```

Fixed. The problem was that not all errors were handled equally. We now also catch errors when calling the controller's `init` function and shut down the script engine when that fails. Otherwise you can get hundreds of script error popups on every slider move.

Holzhaus committed:
controllers/controllerengine: Use key without backtrace for "ignore error"

Be-ing commented on src/controllers/controllerengine.cpp:
```diff
@@ -76,6 +76,8 @@ Output:  -
 void ControllerEngine::callFunctionOnObjects(QList<QString> scriptFunctionPrefixes,
                                              const QString& function, QScriptValueList args) {
+    // This function is called from outside the controller engine, so we can't
+    // use VERIFY_OR_DEBUG_ASSERT here
```
VERIFY_OR_DEBUG_ASSERT can still be used for public functions to verify that other classes are using this correctly. But I'm okay with leaving it as is to avoid more merge conflicts with #1795

Be-ing replied:
Actually this function is private, so I don't understand what the problem with VERIFY_OR_DEBUG_ASSERT would be here.

wrapFunctionCode is called from MidiController though, which doesn't check the state of whether the QScriptEngine exists (confusingly, ControllerEngine has a private `m_pEngine` member which is a QScriptEngine and Controller has a private `m_pEngine` member which is a ControllerEngine. In #1795 we renamed the member in ControllerEngine to `m_pScriptEngine` so don't bother changing that in this PR)

Holzhaus replied:
So there's nothing else to do here?

Be-ing replied:
Use VERIFY_OR_DEBUG_ASSERT here, but leave wrapFunctionCode as it is.

Holzhaus replied:
Done.

Holzhaus committed:
controllers/controllerengine: Use debug assert in callFunctionOnObjects
